### PR TITLE
[feature] Implemented Heap Data Structure #70

### LINKED ISSUE #70: Implement Heap Data Structure
**Is your request related to a problem? Please describe.**

New Feature: Implementing Heap data structure to the Library  

**Describe the solution you'd like**

Use class and define all operations as methods  

**Additional context**

- Add a parameter `min` or `max` to identify min heap or max heap
- Define all possible operations on Heap

### PR DESCRIPTION
**Pull request by 17sushmita**
### Description
Implemented Heap Data Structure
Basic Heap Operations Added:
- Both minHeap or maxHeap supported
- insert, delete, display, to_list functions added

### Related Issue

Fixes #70 

### Type of Change:

- Code
- Quality Assurance
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)

### How Has This Been Tested?

Added heap.ipynb in dsalgo/test to test the operations

### Checklist:

- [x] I read the **CONTRIBUTING** document before making this pull request
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] My change requires change to documentation
- [x] My changes generate no new warnings 
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by ravi5175** [comment]
@17sushmita please follow pep8 guidelines for code formating
and please generate unittest module for heap instead of using jupyter

File: dsalgo/heap.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+class Heap:
+
```
Comment (ravi5175) [inline dsalgo/heap.py:2]: please follow pep8 guidelines of code formatting

File: tests/heap.ipynb [hunk 2]
```diff
@@ -0,0 +1,314 @@
+{
```
Comment (ravi5175) [inline tests/heap.ipynb:1]: please replace heap.ipynb to unittest module for heap

**Review by ravi5175 — CHANGES_REQUESTED** [review]

**Comment by 17sushmita** [comment]
I've followed pep8 guidelines. Will add unittest module

**Comment by ravi5175** [comment]
> I've followed pep8 guidelines. Will add unittest module

please use linter that supports pep8 since i am seeing multiple mistakes in formatting

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Also, follows @17sushmita advice

**Comment by devkapilbansal** [comment]
Also, in last commit change it as:
```
[docs] Updated Readme

Long Description
```

**Comment by 17sushmita** [comment]
@devkapilbansal Please review the changes

File: README.rst [hunk 3]
```diff
@@ -1,13 +1,37 @@
 .. image:: dsalgo.png
   :alt: image not found
 
-**OBJECTIVE:**
```
Comment (devkapilbansal) [inline README.rst:4]: Wouldn't it be better if we add these banners like thing beforer image??🤔

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: tests/heap_tests.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    def test_root(self):
+        for i in self.test_array:
+            self.minHeap.insert(i)
+            self.maxHeap.insert(i)
+        self.assertEqual(self.minHeap.root(), min(self.minHeap.to_list()))
+        self.assertEqual(self.maxHeap.root(), max(self.maxHeap.to_list()))
+
+
+if __name__ == '__main__':
```
Comment (devkapilbansal) [inline tests/heap_tests.py:22]: No need for these lines

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: tests/heap_tests.py (hunk 4 above)
Comment (17sushmita) [reply tests/heap_tests.py:22]: > No need for these lines

As per flake8,  2 blank lines are expected after class or function definition, that's why I left these blank lines

**Comment by ravi5175** [comment]
@17sushmita squash your commits

**Comment by 17sushmita** [comment]
@ravi5175 @devkapilbansal  What changes are required now?

**Comment by ravi5175** [comment]
Squash your commits into two, one for updated readme while other for implementation of heap and heap unittest support

**Comment by 17sushmita** [comment]
@ravi5175 I've squashed the recent commits in 3 because there is a commit for heap in between two readme commits, the readme commits cannot be squashed to one.

**Comment by ravi5175** [comment]
@17sushmita then its okay from my side, approving it

**Review by ravi5175 — APPROVED** [review]

**Comment by ravi5175** [comment]
@17sushmita  Sorry , i forgot i am no longer authorized to approve, please don't tag me for anything

**Comment by devkapilbansal** [comment]
@17sushmita merge your readme changes in one. It is possible.

**Comment by 17sushmita** [comment]
Oh Thanks!! I did it with 
```
git rebase -i HEAD~3
```
and placed the last commit above heap commit

File: tests/heap_tests.py (hunk 4 above)
Comment (devkapilbansal) [reply tests/heap_tests.py:22]: Ohh I am asking for if name=='__main__':
Unittest.main()

File: README.rst [hunk 5]
```diff
@@ -1,13 +1,41 @@
...
+    :alt: issues
+    :scale: 100%
+    :target: https://github.com/codesankalp/dsalgo/issues   
+    
+|img|  
+    
+.. |img| image:: dsalgo.png
+    :alt: image not found
+    :scale: 100%
+
```
Comment (devkapilbansal) [inline README.rst:29]: Why so many lines here?? Are they necessary??

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
See my final request below

File: README.rst (hunk 5 above)
Comment (17sushmita) [reply README.rst:29]: Changes Updated

**Review by devkapilbansal — APPROVED** [review]
LGTM 👍

File: tests/heap.ipynb (hunk 2 above)
Comment (codesankalp) [reply tests/heap.ipynb:1]: I think this file should be removed and all the tests must be mentioned in the unit test file.

**Review by codesankalp — CHANGES_REQUESTED** [review]
please remove `heap.ipynb` and do all the tests in `heap_tests.py`

**Pull request merged**